File: ueueo-framework/ueueo-blob-storing/src/main/java/com/ueueo/BlobStoring/BlobContainer.cs
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Volo.Abp.MultiTenancy;
using Volo.Abp.Threading;

namespace Volo.Abp.BlobStoring;

public class BlobContainer<TContainer> : IBlobContainer<TContainer>
    //where TContainer : class
{
    private readonly IBlobContainer _container;

    public BlobContainer(IBlobContainerFactory blobContainerFactory)
    {
        _container = blobContainerFactory.Create<TContainer>();
    }

    public void SaveAsync(
        String name,
        Stream stream,
        boolean overrideExisting = false,
        CancellationToken cancellationToken = default)
    {
        return _container.SaveAsync(
            name,
            stream,
            overrideExisting,
            cancellationToken
        );
    }

    public Boolean  DeleteAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        return _container.DeleteAsync(
            name,
            cancellationToken
        );
    }

    public Boolean  ExistsAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        return _container.ExistsAsync(
            name,
            cancellationToken
        );
    }

    public Stream GetAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        return _container.GetAsync(
            name,
            cancellationToken
        );
    }

    public Stream GetOrNullAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        return _container.GetOrNullAsync(
            name,
            cancellationToken
        );
    }
}

public class BlobContainer : IBlobContainer
{
    protected String ContainerName;//  { get; }

    protected BlobContainerConfiguration Configuration;//  { get; }

    protected IBlobProvider Provider;//  { get; }

    protected ICurrentTenant CurrentTenant;//  { get; }

    protected ICancellationTokenProvider CancellationTokenProvider;//  { get; }

    protected IServiceProvider ServiceProvider;//  { get; }

    protected IBlobNormalizeNamingService BlobNormalizeNamingService;//  { get; }

    public BlobContainer(
        String containerName,
        BlobContainerConfiguration configuration,
        IBlobProvider provider,
        ICurrentTenant currentTenant,
        ICancellationTokenProvider cancellationTokenProvider,
        IBlobNormalizeNamingService blobNormalizeNamingService,
        IServiceProvider serviceProvider)
    {
        ContainerName = containerName;
        Configuration = configuration;
        Provider = provider;
        CurrentTenant = currentTenant;
        CancellationTokenProvider = cancellationTokenProvider;
        BlobNormalizeNamingService = blobNormalizeNamingService;
        ServiceProvider = serviceProvider;
    }

    public   void SaveAsync(
        String name,
        Stream stream,
        boolean overrideExisting = false,
        CancellationToken cancellationToken = default)
    {
        using (CurrentTenant.Change(GetTenantIdOrNull()))
        {
            var blobNormalizeNaming = BlobNormalizeNamingService.NormalizeNaming(Configuration, ContainerName, name);

            Provider.SaveAsync(
                new BlobProviderSaveArgs(
                    blobNormalizeNaming.ContainerName,
                    Configuration,
                    blobNormalizeNaming.BlobName,
                    stream,
                    overrideExisting,
                    CancellationTokenProvider.FallbackToProvider(cancellationToken)
                )
            );
        }
    }

    public    Boolean  DeleteAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        using (CurrentTenant.Change(GetTenantIdOrNull()))
        {
            var blobNormalizeNaming =
                BlobNormalizeNamingService.NormalizeNaming(Configuration, ContainerName, name);

            return Provider.DeleteAsync(
                new BlobProviderDeleteArgs(
                    blobNormalizeNaming.ContainerName,
                    Configuration,
                    blobNormalizeNaming.BlobName,
                    CancellationTokenProvider.FallbackToProvider(cancellationToken)
                )
            );
        }
    }

    public    Boolean  ExistsAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        using (CurrentTenant.Change(GetTenantIdOrNull()))
        {
            var blobNormalizeNaming =
                BlobNormalizeNamingService.NormalizeNaming(Configuration, ContainerName, name);

            return Provider.ExistsAsync(
                new BlobProviderExistsArgs(
                    blobNormalizeNaming.ContainerName,
                    Configuration,
                    blobNormalizeNaming.BlobName,
                    CancellationTokenProvider.FallbackToProvider(cancellationToken)
                )
            );
        }
    }

    public    Stream GetAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        var stream = GetOrNullAsync(name, cancellationToken);

        if (stream == null)
        {
            //TODO: Consider to throw some type of "not found" exception and handle on the HTTP status side
            throw new AbpException(
                $"Could not found the requested BLOB '{name}' in the container '{ContainerName}'!");
        }

        return stream;
    }

    public    Stream GetOrNullAsync(
        String name,
        CancellationToken cancellationToken = default)
    {
        using (CurrentTenant.Change(GetTenantIdOrNull()))
        {
            var blobNormalizeNaming =
                BlobNormalizeNamingService.NormalizeNaming(Configuration, ContainerName, name);

            return Provider.GetOrNullAsync(
                new BlobProviderGetArgs(
                    blobNormalizeNaming.ContainerName,
                    Configuration,
                    blobNormalizeNaming.BlobName,
                    CancellationTokenProvider.FallbackToProvider(cancellationToken)
                )
            );
        }
    }

    protected   ID GetTenantIdOrNull()
    {
        if (!Configuration.IsMultiTenant)
        {
            return null;
        }

        return CurrentTenant.Id;
    }
}

File: ueueo-framework/ueueo-blob-storing/src/main/java/com/ueueo/blobstoring/BlobContainer.java
package com.ueueo.blobstoring;

import com.ueueo.SystemException;
import com.ueueo.threading.CancellationToken;
import com.ueueo.threading.ICancellationTokenProvider;

import java.io.InputStream;

public class BlobContainer implements IBlobContainer {
    protected String containerName;

    protected BlobContainerConfiguration configuration;

    protected IBlobProvider provider;

    protected ICancellationTokenProvider cancellationTokenProvider;

    protected IBlobNormalizeNamingService blobNormalizeNamingService;

    public BlobContainer(
            String containerName,
            BlobContainerConfiguration configuration,
            IBlobProvider provider,
            ICancellationTokenProvider cancellationTokenProvider,
            IBlobNormalizeNamingService blobNormalizeNamingService
    ) {
        this.containerName = containerName;
        this.configuration = configuration;
        this.provider = provider;
        this.cancellationTokenProvider = cancellationTokenProvider;
        this.blobNormalizeNamingService = blobNormalizeNamingService;

    }

    @Override
    public void save(String name, InputStream stream, boolean overrideExisting, CancellationToken cancellationToken) {
        BlobNormalizeNaming blobNormalizeNaming = blobNormalizeNamingService.normalizeNaming(configuration, containerName, name);

        provider.save(
                new BlobProviderSaveArgs(
                        blobNormalizeNaming.getContainerName(),
                        configuration,
                        blobNormalizeNaming.getBlobName(),
                        stream,
                        overrideExisting,
                        ICancellationTokenProvider.Extensions.fallbackToProvider(cancellationTokenProvider, cancellationToken)
                )
        );
    }

    @Override
    public Boolean delete(String name, CancellationToken cancellationToken) {
        BlobNormalizeNaming blobNormalizeNaming = blobNormalizeNamingService.normalizeNaming(configuration, containerName, name);

        return provider.delete(
                new BlobProviderDeleteArgs(
                        blobNormalizeNaming.getContainerName(),
                        configuration,
                        blobNormalizeNaming.getBlobName(),
                        ICancellationTokenProvider.Extensions.fallbackToProvider(cancellationTokenProvider, cancellationToken)
                )
        );
    }

    @Override
    public Boolean exists(String name, CancellationToken cancellationToken) {
        BlobNormalizeNaming blobNormalizeNaming = blobNormalizeNamingService.normalizeNaming(configuration, containerName, name);

        return provider.exists(
                new BlobProviderExistsArgs(
                        blobNormalizeNaming.getContainerName(),
                        configuration,
                        blobNormalizeNaming.getBlobName(),
                        ICancellationTokenProvider.Extensions.fallbackToProvider(cancellationTokenProvider, cancellationToken)
                )
        );
    }

    @Override
    public InputStream get(String name, CancellationToken cancellationToken) {
        InputStream stream = getOrNull(name, cancellationToken);

        if (stream == null) {
            //TODO: Consider to throw some type of "not found" exception and handle on the HTTP status side
            throw new SystemException(
                    String.format("Could not found the requested BLOB '%s' in the container '%s'!", name, containerName));
        }

        return stream;
    }

    public InputStream getOrNull(String name, CancellationToken cancellationToken) {
        BlobNormalizeNaming blobNormalizeNaming =
                blobNormalizeNamingService.normalizeNaming(configuration, containerName, name);

        return provider.getOrNull(
                new BlobProviderGetArgs(
                        blobNormalizeNaming.getContainerName(),
                        configuration,
                        blobNormalizeNaming.getBlobName(),
                        ICancellationTokenProvider.Extensions.fallbackToProvider(cancellationTokenProvider, cancellationToken)
                )
        );
    }

    public String getContainerName() {
        return containerName;
    }

    public BlobContainerConfiguration getConfiguration() {
        return configuration;
    }

    public IBlobProvider getProvider() {
        return provider;
    }

    public ICancellationTokenProvider getCancellationTokenProvider() {
        return cancellationTokenProvider;
    }

    public IBlobNormalizeNamingService getBlobNormalizeNamingService() {
        return blobNormalizeNamingService;
    }

}
